src/Shared/HttpExtensions.cs

```diff
@@ -1,30 +1,28 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
+using System.Diagnostics.CodeAnalysis;
 using Microsoft.AspNetCore.Http;
 
 internal static class HttpExtensions
 {
-    internal const string UrlEncodedFormContentType = "application/x-www-form-urlencoded";
-    internal const string MultipartFormContentType = "multipart/form-data";
+    private const string UrlEncodedFormContentType = "application/x-www-form-urlencoded";
+    private const string MultipartFormContentType = "multipart/form-data";
 
     internal static bool IsValidHttpMethodForForm(string method) =>
         HttpMethods.IsPost(method) || HttpMethods.IsPut(method) || HttpMethods.IsPatch(method);
 
-    internal static bool IsValidContentTypeForForm(string? contentType)
-    {
-        if (contentType == null)
-        {
-            return false;
-        }
-
-        // Abort early if this doesn't look like it could be a form-related content-type
+    internal static bool HasAnyFormContentType([NotNullWhen(true)] string? contentType)
+        => HasApplicationFormContentType(contentType) || HasMultipartFormContentType(contentType);
 
-        if (contentType.Length < MultipartFormContentType.Length)
-        {
-            return false;
-        }
+    internal static bool HasApplicationFormContentType([NotNullWhen(true)] string? contentType)
+    {
+        // Content-Type: application/x-www-form-urlencoded; charset=utf-8
+        return contentType != null && contentType.Contains(UrlEncodedFormContentType, StringComparison.OrdinalIgnoreCase);
```

Comment:
This is less precise / robust than the prior implementation. It's ok in the normal scenario, but it can cause strange random bugs for other inputs. E.g. this now matches "web-application/x-www-form-urlencoded-foo".

We have tools that can get the same precision as before without allocating:
https://github.com/dotnet/aspnetcore/blob/main/src/Shared/MediaType/ReadOnlyMediaTypeHeaderValue.cs

Reply:
> We have tools that can get the same precision as before without allocating:

Ah, perfect! This is exactly what I was looking for.